2021 年最值得使用的 Node.js 框架 (#8472)
xitu/gold-miner: article/2021/top-node-js-frameworks-to-use-in-2021.md

> * 原文作者：[Ronak Patel](https://medium.com/@ronakataglowid)
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/top-node-js-frameworks-to-use-in-2021.md](https://github.com/xitu/gold-miner/blob/master/article/2021/top-node-js-frameworks-to-use-in-2021.md)
> * 译者：[洛竹](https://github.com/youngjuning)
> * 校对者：

# 2021 年最值得使用的 Node.js 框架

![Top Node.js Frameworks](https://cdn-images-1.medium.com/max/2400/1*9lmHM-CypCHEOtNd_dfRRw.png)

Node.js 是最敏捷的服务端 web 应用平台，因为它为应用开发公司提供了构建可扩展的单一编程语言 web 平台的便利。它是它是最热门的 JavaScript 运行时框架之一，它是开源的，具有跨平台属性，可以在浏览器外创建代码。

## Node.js 的特别之处是什么？

1. 它有一个干净简洁的代码库。
2. 它最适合敏捷开发和原型设计。
3. 它有一个庞大的生态系统，提供开源库。
4. 它可以用于更快地开发企业级可扩展的应用程序。

> 5. 它基于最常用的编程语言--JavaScript。

Reviewer: ```suggestion
5. 它基于最常用的编程语言 JavaScript。
```
```suggestion
5. 它基于最常用的编程语言 —— JavaScript。
```

## 市场对 Nodejs 的反应如何？

![NPM Trends](https://cdn-images-1.medium.com/max/2366/1*4kohl3kB8tfsqcJY-f3xlA.png)

**下面是一些关于 Node.js 的更多信息：**

* Amazon、Netflix、LinkedIn、eBay、PayPal 和 Reddit 使用 Node.js 作为他们的后端框架。
* 43% 的 Node.js 开发者使用这个框架创建企业级应用
* Paypal 注意到改用 Node.js 后，平均响应时间显著缩短了 35%。
* Node.js 可以将任何现有应用的性能提高 50%。
* Node.js 降低了 58% 的应用开发成本。
* Statista 提到，截至 2020 年初，Node 是开发者中使用最多的框架，占 51.4%。

## 使用 Node.js 框架的收益

* 易扩展性
* 学习曲线低
* JavaScript 全栈开发
* 庞大而活跃的社区
* 使用单一的代码库，以提高整体网络性能。

## 2021 年最值得用的 Node.js 框架

我们已经介绍了 Node.js，并详细地了解了它的功能，现在我们可以讨论 2021 年最值得使用的 Node.js 框架啦。

### 1. Hapi.js

- **GitHub**
  - **Stars**：13.1k
  - **Forks**：1.3k
  - **Version**：v20.1.0
  - **Contributors**：208
  - **Used by**：16.5k
- **NPM Weekly Downloads**：105,065

> - **License**：BSD-3-Clause

Reviewer: missing translation

> ![Popularity of Hapi.js framework](https://cdn-images-1.medium.com/max/2316/1*bnq7cugj6hQYaLYlY2a53w.png)

Reviewer: alt

> - **License**：BSD-3-Clause

Author: 原文格式不方便阅读

### 2. Express.js

- **GitHub**
  - **Stars**：52.3k
  - **Forks**：8.8k
  - **Version**：v4.17.1
  - **Contributors**：262
  - **Used by**：9.2m
- **NPM Weekly Downloads**：17,193,915

> - **License**：MIT

Reviewer: missing translation

> ![Express.js popularity](https://cdn-images-1.medium.com/max/2080/1*oV-DuccRP7WvxgSabkTNqQ.png)

Reviewer: alt

> Express.js 是一个灵活而简约的 Node.js 应用框架。这个插件并不是围绕着特定的组件构建的，因此它并不限制你使用什么技术。这就给了开发者尝试的自由。他们还可以获得闪电般的配置和纯JavaScript 体验，这些特性使 Express.js 成为快速原型设计和敏捷开发市场的有力竞争者。

Reviewer: ```suggestion
Express.js 是一个灵活而简约的 Node.js 应用框架。这个插件并不是围绕着特定的组件构建的，因此它并不限制你使用什么技术。这就给了开发者尝试的自由。他们还可以获得闪电般的配置和纯 JavaScript 体验，这些特性使 Express.js 成为快速原型设计和敏捷开发市场的有力竞争者。
```

### 3. Nest.js

- **Github**
  - **Stars**：35.5k
  - **Forks**：3.4k
  - **Version**：7.5.0
  - **Contributors**：226
  - **Used by**：60.4k
- **NPM Weekly Downloads**：508,214

> - **License**：MIT

Reviewer: missing translation

> ![Nest.js popularity](https://cdn-images-1.medium.com/max/2000/1*uTZrckIaqRi2BolefxaVTg.png)

Reviewer: alt

Nest.js 是一个服务器端应用框架，它是为了解放开发者的生产力，让他们的生活变得更轻松而打造的。开发者通常为了更好地组织和管理代码而使用这个 Node.js 框架。

**Nest.js 可以被用于：**

* 编写更清晰和可重用的代码。
* 编写具有更高层次结构的代码，如拦截器、过滤器、管道等；
* 编写可扩展、可测试和松散型应用程序。

**Nest.js 主要特性**

* 易于扩展：可与其他库一起使用。
* 允许开发人员完全使用纯 JavaScript 进行编码。

> * 结合了函数式编程、面向对象编程和反应式编程的特点。

Reviewer 2: 反应式编程 => 函数响应式编程

### 4. Koa.js

- **GitHub**
  - **Stars**：30.9k
  - **Forks**：3k

>   - **Version**：2.13.1
>   - **Contributors**：219
>   - **Used by**：160k
> - **NPM Weekly Download**：870,944
> - **License**：MIT

Reviewer: ```suggestion
  - **Version**：2.13.1
  - **Contributors**：219
  - **Used by**：160k
- **NPM Weekly Download**：870,944
- **License**：MIT
```

> ![Popularity of Koa.js framework](https://cdn-images-1.medium.com/max/2078/1*lKpRVNszZ6gVbTF_blVJRQ.png)

Reviewer: alt

> Koa.js 是一个开源的 Node 网络框架，由 Express.js 原班人马创建。通过 Koa，他们的目标是为 Web 应用和 API 创建一个更小、更有价值、更强大的平台。它提供了多种高效的方法，以让构建服务的过程更快速。

Reviewer 2: 网络框架 => web 框架

### 5. Socket.io

- **GitHub**
  - **Stars**：52.7k
  - **Forks**：9.6k

>   - **Version**：4.0.0
>   - **Contributors**：189
>   - **Used by**：2.2m
> - **NPM Weekly Downloads**：3,617,636
> - **License**：MIT

Reviewer: ```suggestion
  - **Version**：4.0.0
  - **Contributors**：189
  - **Used by**：2.2m
- **NPM Weekly Downloads**：3,617,636
- **License**：MIT
```

> ![Popularity of socket.io](https://cdn-images-1.medium.com/max/2046/1*5zUKf_Dt48W00aWfkAs84g.png)

Reviewer: alt

Socket.io 是用来在客户端和服务器端之间创建实时双向通信的框架。

> 要做到这一点，客户端需要在浏览器中安装 Socket.io，服务器也要集成 Socket.io 包。这使得数据可以在数百万种形式中共享。然而，最受欢迎的方法仍然是 JSON。

Reviewer 2: 段落错位了

> 要做到这一点，客户端需要在浏览器中安装 Socket.io，服务器也要集成 Socket.io 包。这使得数据可以在数百万种形式中共享。然而，最受欢迎的方法仍然是 JSON。

Reviewer 2: 可以在数百万种形式中共享 => 可以以各种形式共享

Socket.io is used for establishing real-time bi-directional communication between clients and servers. For this to work, the client needs to have Socket.io in their browser, and the server should have also integrated the Socket.io package. This allows data to be shared in millions of forms. However, the most preferred method remains JSON.
Socket.io 是用来在客户端和服务器端之间创建实时双向通信的框架。

**Socket.io consists of**
要做到这一点，客户端需要在浏览器中安装 Socket.io，服务器也要集成 Socket.io 包。这使得数据可以在数百万种形式中共享。然而，最受欢迎的方法仍然是 JSON。

1. JavaScript Server — Node.js
**Socket.io 由以下两个部分组成：**

2. JavaScript client library — Node.js
1. JavaScript 服务：Node.js
2. JavaScript 客户端库：Node.js

> **Note:** Socket.io is also compatible with many other languages like Java, C+, Swift, Dart, .Net, and Python.
> **注意：** Socket.io 还兼容许多其他语言，如 Java、C+、Swift、Dart、.Net 和 Python。

**Socket.io can be used for -**
**Socket.io 可以被用于：**

* Several namespaces
* Broadcasting
* Event Handling
* Error Handling
* Logging and Debugging

> * Internal

Reviewer 2: Internal 漏了

> * 速度和可靠性突出

Reviewer 2: 速度和可靠性突出 => 出色的速度和可靠性

> Socket.io 是最好的基于事件的实时双向通信工具之一。任何想要在应用中添加实时分析功能的人都应该使用它。Socket.io 对于实时游戏应用也很有用。在实时游戏中使用基本的 HTTP 或 HTTPS 协议是不可行的，因为这些文件很大，建立通信需要时间。在这里，我们使用体积更小的 socket 包，几乎是实时地完成工作，以获得更流畅和更好的体验。

Reviewer 2: 我们使用体积更小的 socket 包，几乎是实时地完成工作  => 我们使用体积更小的 socket 包来保证准“实时”

### 6. Meteor.js

- **Github**
  — **Stars**：42.3k
  - **Forks**：5.2k

>   - **Version**：4.0.0
>   - **Contributors**：452
> - NPM：不可用
> - **License**：MIT

Reviewer: ```suggestion
  - **Version**：4.0.0
  - **Contributors**：452
- NPM：不可用
- **License**：MIT
```

Meteor.js 是一个开源的全栈 JavaScript 平台，JavaScript 根据意图不同运行在不同的地方。JavaScript 运行在 Web 浏览器内部；然后 JavaScript 运行在 Node.js 容器内的 Meteor 服务器上，支持 HTML 片段、静态资源和 CSS 规则。

**Meteor.js 可以被用于：**

* 移动应用程序全流程
* 网络应用程序全流程

**Meteor.js 主要特性：**

* 纯 JavaScript
* 干净、稳健的数据同步

> * 互操作性

Reviewer 2: 互操作性 => 互通性

### 7. Adonis.js

- **Github**
  - **Stars**：9.7k
  - **Forks**：498
  - **版本**：5.0.13
  - **贡献者**：47
  - **使用人数**：754

> - **NPM Weekly Downloads**：3808

Reviewer 2: 漏翻周下载数

> - **License**：MIT

Reviewer: missing translation

> ![Popularity of adoni.js](https://cdn-images-1.medium.com/max/2052/1*QVu2Z2wbudWv06tBwkJ8mQ.png)

Reviewer: alt

**NPM — Weekly Downloads** - 3808 **| License** - MIT
![adoni.js 流行度](https://cdn-images-1.medium.com/max/2052/1*QVu2Z2wbudWv06tBwkJ8mQ.png)

![Popularity of adoni.js](https://cdn-images-1.medium.com/max/2052/1*QVu2Z2wbudWv06tBwkJ8mQ.png)
Adonis.js 是一个 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) 的 MVC 框架，可以运行在所有的操作系统上。它为编写服务器端 Web 应用程序提供了一个稳定的生态系统，以让开发者专注于业务需求，如最终确定选择或排除哪个包。

> Adonis.js is a [Node.js ](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/)MVC framework that can run on all operating systems. It provides a stable ecosystem for writing server-side web applications to focus on the business requirements like finalizing which package to choose or exclude. It is ideal for Laravel Developers who want a taste change and are trying out the [](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/)[Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) framework. Adonis.js provides [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) with the same features and abilities that Laravel has naturally.

Reviewer 2: 漏翻
Adonis.js provides [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) with the same features and abilities that Laravel has naturally.

* 构建 web 应用
* 应用程序接口服务

**Adonis.js 主要特性：**

* 强大的 ORM，帮助进行安全的 SQL 查询。
* API 和基于会话的认证系统

> * 验证和给每一个用户的输入做卫生处理。

Reviewer 2: 验证和给每一个用户的输入做卫生处理 => 验证和清理用户输入

### 8. Sails.js

- **Github**
  - **Stars**：21.9k
  - **Forks**：1.9k
  - **版本**：1.4.0
  - **贡献者**：229
  - **使用人数**：23.3k

> - **NPM Weekly Downloads**：20,457

Reviewer 2: 漏翻周下载数

> - **License**：MIT

Reviewer: missing translation

> ![Popularity of sails.js framework](https://cdn-images-1.medium.com/max/2000/1*nYu5aJOd4f933c6WZkX8eA.png)

Reviewer: alt

> Sails.js是又一个实时 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) MVC框架。它基于 Express 构建，其 MVC 架构与 Ruby on Rails 相似。它与 Ruby on Rails 的不同之处在于，它提供了对更现代的、以数据为中心的 API 和 Web 应用开发风格的支持。

Reviewer: ```suggestion
Sails.js 是又一个实时 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) MVC 框架。它基于 Express 构建，其 MVC 架构与 Ruby on Rails 相似。它与 Ruby on Rails 的不同之处在于，它提供了对更现代的、以数据为中心的 API 和 Web 应用开发风格的支持。
```

**Sails.js 可以被用于：**

* 构建企业级 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) 应用
* 构建前端应用
* 构建处理 HTTP 请求的后端应用

**Sails.js 主要特性：**

* 支持自动生成 REST APIs
* 具有简单的 WebSocket 集成
* 兼容任何流行的前端框架：Angular、Android、React、iOS、Windows 或任何自定义硬件。
* 它还具有实时功能支持

**什么时候使用 Sails.js：**

> 任何想要一个模拟 MVC 模式的 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) 框架（如Laravel和Ruby on Rails）、想要实现现代应用架构，并构建以数据为中心的 API 和实时应用的开发者都应该在他们的下一个项目中使用 Sails.js。

Reviewer: ```suggestion
任何想要一个模拟 MVC 模式的 [Node.js](https://aglowiditsolutions.com/blog/hidden-cost-of-app-development/) 框架（如 Laravel 和 Ruby on Rails）、想要实现现代应用架构，并构建以数据为中心的 API 和实时应用的开发者都应该在他们的下一个项目中使用 Sails.js。
```
